plivo.py

```diff
@@ -20,12 +21,37 @@ class PlivoError(Exception):
     pass
 
 
-def validate_signature(uri, post_params, signature, auth_token):
-    for k, v in sorted(post_params.items()):
-        uri += k + v
-    return base64.encodestring(hmac.new(auth_token, uri, sha1).digest()).strip() == signature
-
-
+def validate_request_signature(uri, signature, auth_token, params=None):
+    """
+    Validates requests made by Plivo to your servers.
+    See https://www.plivo.com/docs/xml/request/#validation
+
+    :param uri: Your server URL
+    :param params: POST Parameters passed to your URL, in case of POST request. Will be ignored if URL contains a query
+    string
+    :param signature: X-Plivo-Signature header
+    :param auth_token: Plivo Auth token
+    :return: True if the request matches signature, False otherwise
+    """
+    parsed_uri = urlparse(uri.encode('utf-8'))
+    qs = parsed_uri.query
+    if qs:
+        # get params from query string
+        all_params = dict(parse_qsl(qs, keep_blank_values=True))
+        # remove parameters from query string
+        encoded_request = urljoin(uri, parsed_uri.path).encode('utf-8')
+    else:
+        all_params = params.copy() if params else dict()
```

Comment:
You're not altering `params`, so there's no point to copy them. I think it's much cleaner to write it as: `all_params = params or {}`.